src/module_compilation.ml

```diff
@@ -48,16 +51,23 @@ let build_cm sctx ?sandbox ~dynlink ~flags ~cm_kind ~(dep_graph:Ocamldep.dep_gra
                deps
                ~f:(fun m ->
                  match cm_kind with
-                 | Cmi | Cmo -> [Module.cm_file m ~dir Cmi]
-                 | Cmx -> [Module.cm_file m ~dir Cmi; Module.cm_file m ~dir Cmx])))
+                 | Cmi | Cmo -> [Module.cm_file m ~obj_dir Cmi]
+                 | Cmx -> [Module.cm_file m ~obj_dir Cmi; Module.cm_file m ~obj_dir Cmx])))
       in
-      let extra_targets, cmt_args =
+      let extra_targets, extra_symlinks, cmt_args =
         match cm_kind with
-        | Cmx -> (extra_targets, Arg_spec.S [])
+        | Cmx -> (extra_targets, extra_symlinks, Arg_spec.S [])
         | Cmi | Cmo ->
-          let fn = Option.value_exn (Module.cmt_file m ~dir ml_kind) in
-          (fn :: extra_targets, A "-bin-annot")
+          let fn = Option.value_exn (Module.cmt_file m ~obj_dir ml_kind) in
+          let fn2 = Option.value_exn (Module.cmt_file m ~obj_dir:dir ml_kind) in
+          (fn :: extra_targets, fn2 :: extra_symlinks, A "-bin-annot")
       in
+      let old_dst = Module.cm_file m ~obj_dir:dir cm_kind in
+      if obj_dir <> dir then begin
+        SC.add_rule sctx ?sandbox (Build.symlink ~src:dst ~dst:old_dst) ;
+        List.iter2 extra_targets extra_symlinks ~f:(fun src dst ->
+          SC.add_rule sctx ?sandbox (Build.symlink ~src ~dst))
```

Comment:
Instead of carefully building the two lists symmetrically, what about just building the list of targets as a string list (it would just be the filenames), then at the end we do:

```
let targets  = List.map target_filenames ~f:(Path.relative ~dir:obj_dir) in
let symlinks = List.map target_filenames ~f:(Path.relative ~dir) in
...
```